receiver/sapmreceiver/factory.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2019, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sapmreceiver
+
+// This file implements factory for Jaeger receiver.
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strconv"
+
+	"github.com/open-telemetry/opentelemetry-collector/config/configerror"
+	"github.com/open-telemetry/opentelemetry-collector/config/configmodels"
+	"github.com/open-telemetry/opentelemetry-collector/consumer"
+	"github.com/open-telemetry/opentelemetry-collector/receiver"
+	"go.uber.org/zap"
+)
+
+const (
+	// The value of "type" key in configuration.
+	typeStr = "sapm"
+
+	// Default endpoints to bind to.
+	defaultEndpoint = "localhost:7276"
+)
+
+// Factory is the factory for Jaeger receiver.
+type Factory struct {
+}
+
+// Type gets the type of the Receiver config created by this factory.
+func (f *Factory) Type() string {
+	return typeStr
+}
+
+// CustomUnmarshaler returns nil because we don't need custom unmarshaling for this config.
+func (f *Factory) CustomUnmarshaler() receiver.CustomUnmarshaler {
+	return nil
+}
+
+// CreateDefaultConfig creates the default configuration for Jaeger receiver.
+func (f *Factory) CreateDefaultConfig() configmodels.Receiver {
+	return &Config{
+		ReceiverSettings: configmodels.ReceiverSettings{
+			TypeVal:  typeStr,
+			NameVal:  typeStr,
+			Endpoint: defaultEndpoint,
+		},
+	}
+}
+
+// extract the port number from string in "address:port" format. If the
+// port number cannot be extracted returns an error.
+// TODO make this a utility function
+func extractPortFromEndpoint(endpoint string) (int, error) {
+	_, portStr, err := net.SplitHostPort(endpoint)
+	if err != nil {
+		return 0, fmt.Errorf("endpoint is not formatted correctly: %s", err.Error())
+	}
+	port, err := strconv.ParseInt(portStr, 10, 0)
+	if err != nil {
+		return 0, fmt.Errorf("endpoint port is not a number: %s", err.Error())
+	}
+	if port < 1 || port > 65535 {
+		return 0, fmt.Errorf("port number must be between 1 and 65535")
+	}
+	return int(port), nil
+}
+
+// CreateTraceReceiver creates a trace receiver based on provided config.
+func (f *Factory) CreateTraceReceiver(
+	ctx context.Context,
+	logger *zap.Logger,
+	cfg configmodels.Receiver,
+	nextConsumer consumer.TraceConsumer,
+) (receiver.TraceReceiver, error) {
+	// assert config is SAPM config
+	rCfg := cfg.(*Config)
+
+	port, err := extractPortFromEndpoint(rCfg.Endpoint)
```

Comment:
I copied this check from the jaeger receiver.  I think it's primary purpose is to prevent a 0 valued port from being passed to the listener.  This wouldn't raise an error it would just assign a random port to the listener.

Reply:
hum, perhaps we need a better and general way to handle that. Anyway, it is a separated issue.